### PR TITLE
Hide binder type in Ltac2

### PR DESCRIPTION
**Pull request by ppedrot**
For robustness, it is better to leave the Ltac2 binder type opaque. Users are not supposed to fiddle with it. The internal representation was originally introduced by @SkySkimmer when porting Ltac2 to SProp, but the design didn't receive much thinking. Technically, I believe this is even the API that the kernel should be exposing, rather than the one used currently that was essentially motivated by the fact it required less code change.

It is an API change, so this may break code in the wild. Yet it is better to do so while Ltac2 is still deemed experimental, and furthermore this change makes it easier to make code portable in the future as it hides implementation details.

### CODE REVIEW

File: user-contrib/Ltac2/tac2core.ml [hunk 1]
```diff
@@ -664,6 +635,23 @@ let () = define1 "constr_pretype" (repr_ext val_preterm) begin fun c ->
   pf_apply pretype
 end
 
+let () = define2 "constr_binder_make" (option ident) constr begin fun na ty ->
+  pf_apply begin fun env sigma ->
+  let rel = Retypeops.relevance_of_term env (EConstr.to_constr sigma ty) in
```
Comment (SkySkimmer) [inline user-contrib/Ltac2/tac2core.ml]: Should be _of_type no?

**Comment by tchajed** [comment]
This does break my code, but it's easy to adapt to, and a much better API because as you said it's more future-proof.

**Review by JasonGross — COMMENTED** [review]
Maybe it's worth adding test cases early in Ltac2?  Like, just a minor case to show that things round-trip correctly with binders, say, of type Type, Set, nat, SProp, Prop, and some binder that lives in a type that lives in SProp?

File: user-contrib/Ltac2/tac2core.ml (hunk 1 above)
Comment (SkySkimmer) [reply user-contrib/Ltac2/tac2core.ml]: No  mean
```suggestion
  let rel = Retypeops.relevance_of_type env (EConstr.to_constr sigma ty) in
```

Comment (ppedrot) [reply user-contrib/Ltac2/tac2core.ml]: Ah, right. Will do.

**Comment by ppedrot** [comment]
I have added round-trip tests, this is ready and needs an assignee.

**Comment by JasonGross** [comment]
If some admin approves my request to join @coq/ltac2-maintainers , I can review and assign and merge.

**Comment by ppedrot** [comment]
@JasonGross done. You should read the merge documentation in CONTRIBUTING.md.

**Comment by JasonGross** [comment]
Has it changed recently?  Do I need to do anything other than running `merge-pr.sh`?

**Review by JasonGross — APPROVED** [review]
fiat-parsers failure is spurious

**Comment by ppedrot** [comment]
@JasonGross no, I was just being cautious since the two last persons I gave the merge rights performed a manual merge.

**Comment by tchajed** [comment]
If this lands only on 8.12 and not 8.11.1, my one real-world use case for Ltac2 will not work with both versions. (I need it to convert strings to identifiers and pass the result to Ltac1, by putting the identifier inside a binder.)

**Comment by ppedrot** [comment]
@tchajed this is not backportable in 8.11, because it breaks the API without a strong reason. I recommend writing version-dependent code, unfortunately.

**Comment by JasonGross** [comment]
Is there a way to write version-dependent code in Ltac2 without needing special Makefile rules that select different files with different versions?
Perhaps I should open an issue for getting version pragmas in Coq?

**Pull request merged**